File: 01_make-aws-lambda-handler-testable/nodejs/code_final/src/services.js
const AWS = require('aws-sdk')

exports.bootstrap = function (cost_center) {
    return exports.handler_factory(cost_center);
}

exports.handler_factory = function (cost_center) {
    if (!cost_center) {
        throw new Error("expecting cost center ID")
    }

    const client = new AWS.Lambda();

    return async function (event, context) {
        if (!event.arn) {
            throw new Error("expecting 'arn' key")
        }

        await client.tagResource({
            Resource: event.arn,
            Tags: {
                CostCenter: cost_center
            }
        }).promise()

        response = await client.listTags({
            Resource: event.arn
        }).promise()

        return response.Tags
    }
}

File: 01_make-aws-lambda-handler-testable/python/code_final/src/services.py
from typing import Callable

import boto3


def bootstrap(cost_center: str) -> Callable[[dict, object], dict]:
    session = boto3.Session()

    return handler_factory(session, cost_center)


def handler_factory(
    session: boto3.Session, cost_center: str
) -> Callable[[dict, object], dict]:
    assert cost_center, "expecting cost center ID"

    lambda_client = session.client("lambda")

    def handler(event: dict, context: object) -> dict:
        assert "arn" in event, "expecting 'arn' key"

        lambda_client.tag_resource(
            Resource=event["arn"], Tags={"CostCenter": cost_center},
        )

        response = lambda_client.list_tags(Resource=event["arn"],)

        return response["Tags"]

    return handler
